src/class-wp-directive-processor.php

```diff
@@ -0,0 +1,92 @@
+<?php
+
+class WP_Directive_Processor extends WP_HTML_Tag_Processor {
+	const DIRECTIVE_PREFIX = 'WPX-';
+
+	public $html;
+
+	public $might_have_directives = true;
+
+	public function __construct( $html ) {
+		parent::__construct( $html );
+
+		if ( false === stripos( self::DIRECTIVE_PREFIX, $html ) ) {
+			$this->might_have_directives = false;
+		}
+	}
+
+	public function next_directive() {
+		if ( false === $this->might_have_directives ) {
+			return false;
+		}
+
+		while ( $this->next_tag( array( 'tag_closers' => 'visit' ) ) ) {
+			$tag_name = $this->get_tag();
+			if ( 0 === stripos( self::DIRECTIVE_PREFIX, $tag_name ) ) {
+				return true;
+			}
+
+			$attribute_directives = $this->get_attribute_names_with_prefix( self::DIRECTIVE_PREFIX );
+			if ( 0 < count( $attribute_directives ) ) {
+				return true;
+			}
+		}
+
+		return false;
+	}
+
+	public function next_balanced_closer() {
+		$depth = 0;
+
+		$tag_name = $this->get_tag();
+		while ( $this->next_tag( array( 'tag_name' => $tag_name, 'tag_closers' => 'visit' ) ) ) {
+			if ( ! $this->is_tag_closer() ) {
+				$depth++;
+				continue;
+			}
+
+			if ( 0 === $depth ) {
+				return true;
+			}
+
+			$depth--;
+		}
+
+		return false;
+	}
+
+	public function get_inner_html() {
```

Comment:
It's only `get_inner_html` if we assume normative markup. I'd suggest this:

```suggestion
	public function get_balanced_tags_html() {
```

Alternatively, we could return false in `next_balanced_closer` if it detects any misnesting.

Reply:
Also, the current tag is a void element.

Reply:
thanks @adamziel - while not in the comments (but found in the commit message), because I was tossing this together quite quickly, I was taking this file to assume certain things:

> assuming normative markup generated from React, assuming that all tag directives contain a balanced closer.

so my idea was simply to write it as if that were true and we can treat deviations as bugs. I don't think any of the directives are self-closing foreign elements, though if they are we can add that support.

this is more or less my exploration of a client API where I consider strictness needs to be lower overall. e.g. if we're looking for the inner HTML of a `wp-context` then it won't matter if we have void tags or unbalanced tags. everything inside of the directives is blank since we're jumping straight to the next `wp-context` tag.

Reply:
If the markup is supposed to be normative then there should be no harm in baling out in case misnesting is detected?

Reply:
But then again, assuming everything inside is skippable like a text means there’s no need to do these checks

Reply:
we can't assume the contents are Text/Data. in fact, we _expect_ them to be HTML, which means we still need to scan them to avoid the conflation of things like `<wp-show><textarea>Use <wp-show when="…"> to conditionally show things</textarea></wp-show>`